cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/services/ProgramRunStatusMonitorService.java

```diff
@@ -0,0 +1,109 @@
+
+  private final Store store;
+  private final ProgramRuntimeService runtimeService;
+  private final int txBatchSize;
+  private final long interval;
+
+  @Inject
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService) {
+    this(cConf, store, runtimeService, cConf.getInt(Constants.AppFabric.PROGRAM_TERMINATOR_TX_BATCH_SIZE));
+  }
+
+  @VisibleForTesting
+  ProgramRunStatusMonitorService(CConfiguration cConf, Store store, ProgramRuntimeService runtimeService,
+                                 int txBatchSize) {
+    super(RetryStrategies.fromConfiguration(cConf, Constants.Service.PROGRAM_RUN_STATUS_MONITOR_SERVICE + "."));
+    this.store = store;
+    this.runtimeService = runtimeService;
+    this.txBatchSize = txBatchSize;
+    long specifiedInterval = cConf.getLong(Constants.AppFabric.PROGRAM_TERMINATOR_INTERVAL_MILLIS);
+    if (specifiedInterval <= 0) {
+      LOG.warn("Invalid interval specified for the program terminator {}. Setting it to 5 minutes.",
+               specifiedInterval);
+      specifiedInterval = 30000L;
+    }
+    this.interval = specifiedInterval;
+  }
+
+  @Override
+  protected long runTask() {
+    LOG.debug("Starting ProgramRunStatusMonitorService");
+    terminateStoppingProgramsActiveBeyondTerminateTime();
+    return this.interval;
+  }
+
+  void terminateStoppingProgramsActiveBeyondTerminateTime() {
+    // fetch all runs that are in stopping state that started at most a minute before current time.
+    // Specifying the entire time range should not be worse in performance
+    // than specifying a more restrictive time range because time range is just used as a read-time filter
+    long currentTimeInMillis = System.currentTimeMillis();
+    Map<ProgramRunId, RunRecordDetail> stoppingRuns = store.getRuns(ProgramRunStatus.STOPPING, 0L,
+                                                                    currentTimeInMillis - 60000, txBatchSize, null);
+    for (Map.Entry<ProgramRunId, RunRecordDetail> stoppingRun : stoppingRuns.entrySet()) {
+      RunRecordDetail runRecordDetail = stoppingRun.getValue();
+      Long terminateTime = runRecordDetail.getTerminateTs();
+      // If terminateTime or RuntimeInfo are null, there isn't anything that can be done about it, so we just ignore it
+      if (terminateTime != null && terminateTime <= TimeUnit.MILLISECONDS.toSeconds(currentTimeInMillis)) {
+        ProgramRunId programRunId = runRecordDetail.getProgramRunId();
+        RuntimeInfo runtimeInfo = runtimeService.lookup(programRunId.getParent(),
+                                                        RunIds.fromString(programRunId.getRun()));
+        if (runtimeInfo != null) {
+          ProgramController controller = runtimeInfo.getController();
+          if (controller != null) {
+            LOG.warn("Forcing the termination of program {} as it should have stopped at {} ",
```

Comment:
warnings usually indicate some issue with the system. In this case, appfabric is working as intended, so an info message is more fitting.

Reply:
Makes sense